nixos/nixpkgs: add option to specify Nixpkgs source path

rnhmjoj opened the pull request:
A non-breaking and (IMHO) simpler alternative to https://github.com/NixOS/nixpkgs/pull/333788 for setting the Nixpkgs versions declaratively from configuration.nix.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change (`nixos-rebuild`, `nixos-rebuild-ng`)
- [x] Tested basic functionality of all binary files
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

rnhmjoj committed:
nixos/nixpkgs: add option to specify Nixpkgs source path

This moves the options under `nixpkgs.flake` to `nixpkgs` and
generalises them to work with systems built without flakes.

Specifically, nixpkgs.source now allows to pin the version of nixpkgs
used by nixos-rebuild to build the system, even without flakes.

rnhmjoj committed:
nixos-rebuild: use nixpkgs.source over <nixpkgs> if set

rnhmjoj committed:
doc/release-notes: mention nixpkgs.source

rnhmjoj committed:
nixos-rebuild-ng: use nixpkgs.source over <nixpkgs> if set

thiagokokada commented on pkgs/by-name/ni/nixos-rebuild-ng/src/tests/test_models.py:
```diff
@@ -4,12 +4,21 @@
 
-
-def test_build_attr_from_arg() -> None:
+@patch(get_qualified_name(nr.process.subprocess.run), autospec=True)
+def test_build_attr_from_arg(mock_run: Any, tmp_path: Path) -> None:
+    nixpkgs_path = tmp_path / "nixpkgs"
+    mock_run.side_effect = [
+        # with config.nixpkgs.source
+        subprocess.CompletedProcess([], 0, f"\"{nixpkgs_path}\"\n"),
+        # without config.nixpkgs.source
+        subprocess.CompletedProcess([], 1, ""),
+    ]
+    assert m.BuildAttr.from_arg(None, None) == m.BuildAttr(str(nixpkgs_path / "nixos"), None)
     assert m.BuildAttr.from_arg(None, None) == m.BuildAttr("<nixpkgs/nixos>", None)
```
I think it is better to do a scoped patch here instead:
```suggestion
def test_build_attr_from_arg(tmp_path: Path) -> None:
    nixpkgs_path = tmp_path / "nixpkgs"
    with patch(get_qualified_name(nr.process.subprocess.run), autospec=True, return_value=subprocess.CompletedProcess([], 0, f"\"{nixpkgs_path}\"\n")):
        assert m.BuildAttr.from_arg(None, None) == m.BuildAttr(str(nixpkgs_path / "nixos"), None)
    with patch(get_qualified_name(nr.process.subprocess.run), autospec=True, return_value=subprocess.CompletedProcess([], 1, "")):
    assert m.BuildAttr.from_arg(None, None) == m.BuildAttr("<nixpkgs/nixos>", None)
```

rnhmjoj replied:
Done, thank you!

thiagokokada reviewed (commented):
Can't comment too much about the change itself since I use Flakes, but at least the `nixos-rebuild-ng` changes looks good to me.

Left a comment about the tests mocks but this is not a merge blocker.

amozeo reviewed (changes requested):
Hey, `nixpkgs.source` option in nixos is an interesting approach to pin nixpkgs for a given nixos configuration. I found several issues with your implementation of it.

I cannot review changes in nixos-rebuild-ng because I am not familiar with its code.

in nixos-rebuild.sh, there is also one more instance of getting nixpkgs from nix path in line 646 (after changes). I think you forgot about that occurence.

amozeo commented:
I just noticed that there is option `nixpkgs.flake.sources` that tries to hold similar information as the option that you want to add, `nixpkgs.source`, maybe we want to deprecate it/change it\reuse it? doesn't look like it's used anywhere else in nixos module system.

rnhmjoj commented:
> Hey, nixpkgs.source option in nixos is an interesting approach to pin nixpkgs for a given nixos configuration.
> I found several issues with your implementation of it.

Thank you for reviewing!

> in nixos-rebuild.sh, there is also one more instance of getting nixpkgs from nix path in line 646 (after changes). I think you forgot about that occurence.

Yes, I missed it because I was looking only for the angle brackets syntax.

rnhmjoj commented on nixos/modules/misc/nixpkgs.nix:
```diff
@@ -157,6 +157,30 @@ in
 
+    source = lib.mkOption {
+      type = lib.types.str;
+      defaultText = "<nixpkgs>";
```
Uhm, I'm actually not sure what happens if you use flakes. I guess it will be completely ignored.

rnhmjoj replied:
Done.

amozeo replied:
<details><summary>Resolved</summary>
<p>

I cannot mark this thread as resolved actually on Github, so I'm commenting that this thread, that I created, is for me resolved

</p>
</details>

rnhmjoj commented on nixos/modules/misc/nixpkgs.nix:
```diff
@@ -157,6 +157,30 @@ in
+      description = ''
+        If set, the path to the Nixpkgs source that will be used by the
+        `nixos-rebuild` tool to build the NixOS system, otherwise the default
+        is to use the `<nixpkgs>` channel.
```
Done.

amozeo replied:
<details><summary>Resolved</summary>
<p>

I cannot mark this thread as resolved actually on Github, so I'm commenting that this thread, that I created, is for me resolved

</p>
</details>

rnhmjoj commented on nixos/modules/misc/nixpkgs.nix:
```diff
@@ -157,6 +157,30 @@ in
     };
 
+    source = lib.mkOption {
+      type = lib.types.str;
```
Uhm, to be equivalent to `<nixpkgs>` it should also allow local paths. For example, using a local checkout of Nixpkgs with some patch could be useful. Also `builtins.fetchTarball` does not return a derivation, so that type may be confusing.

amozeo replied:
<details><summary>Resolved</summary>
<p>

I cannot mark this thread as resolved actually on Github, so I'm commenting that this thread, that I created, is for me resolved

</p>
</details>

rnhmjoj commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -489,7 +489,7 @@ if [[ -z $_NIXOS_REBUILD_REEXEC && -n $canRun && -z $fast ]]; then
         SHOULD_REEXEC=1
     elif [[ -z $flake ]]; then
-        if p=$(runCmd nix-build --no-out-link --expr 'with import <nixpkgs/nixos> {}; config.system.build.nixos-rebuild' "${extraBuildFlags[@]}"); then
+        if p=$(runCmd nix-build --no-out-link --expr 'with import "$nixpkgsPath/nixos" {}; config.system.build.nixos-rebuild' "${extraBuildFlags[@]}"); then
```
It should be fixed now.

amozeo replied:
<details><summary>Resolved</summary>
<p>

I cannot mark this thread as resolved actually on Github, so I'm commenting that this thread, that I created, is for me resolved

</p>
</details>

rnhmjoj commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -604,6 +621,8 @@ getVersion() {
 }
 
 
+getNixpkgsPath
```
Done.

amozeo replied:
<details><summary>Resolved</summary>
<p>

I cannot mark this thread as resolved actually on Github, so I'm commenting that this thread, that I created, is for me resolved

</p>
</details>

rnhmjoj commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -553,16 +549,37 @@ prebuiltNix() {
+    NIXOS_CONFIG=${NIXOS_CONFIG:-$(runCmd nix-instantiate --find-file nixos-config)}
+    if [[ -d $NIXOS_CONFIG ]]; then
+        NIXOS_CONFIG=$NIXOS_CONFIG/default.nix
+    fi
```
I think it makes sense now that `getNixpkgsPath` is at the top (what I had intended to do).

amozeo replied:
I see why you moved here, you use it 8 lines further down.

amozeo commented on nixos/modules/misc/nixpkgs.nix:
```diff
@@ -359,6 +383,16 @@ in
+    # Set nixpkgs in NIX_PATH if nix-channel is disabled
+    nix.nixPath = lib.optional (
+      config.nixpkgs ? source && !config.nix.channel.enable
+    ) "nixpkgs=/run/current-system/nixpkgs";
```
> Wouldn't it be garbage collected otherwise?
no, since the nixpkgs defined in `nixpkgs.source` will be a dependency of etc environment.

rnhmjoj replied:
So you're proposing to keep the `nixPath` entry by default?

amozeo replied:
I'm proposing to

1. not symlinking `nixpkgs.source` in system derivation
2. add enable option for behavior of adding `nixpkgs.source` to nix path

rnhmjoj replied:
Ok, done.

amozeo replied:
<details><summary>Resolved</summary>
<p>

I cannot mark this thread as resolved actually on Github, so I'm commenting that this thread, that I created, is for me resolved

</p>
</details>

rnhmjoj commented:
> I just noticed that there is option nixpkgs.flake.sources that tries to hold similar information 

I think we should unify the two options.

rnhmjoj commented:
I made an attempt to unify flakes and non-flakes options in 39b947396df00acb4fc63069fdbd82d3d87a654d. The flake part is completely untested for now.

thiagokokada reviewed (approved):
`nixos-rebuild-ng` part LGTM, but not tested.

Also didn't review the other changes.

amozeo commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -218,6 +218,27 @@ runCmd() {
+    nixpkgsPath=
+
+    # Get NixOS configuration
+    NIXOS_CONFIG=${NIXOS_CONFIG:-$(runCmd nix-instantiate --find-file nixos-config)}
```
nixos-rebuild fails here if `NIXOS_CONFIG` environment variable is not set and `nixos-config` nix path is not set. (This is a case for my system)

rnhmjoj replied:
I thought `nixos-config=/etc/nixos/configuration.nix` was in the default NIX_PATH. Does `nixos-rebuild edit` not work for you?

amozeo replied:
It didn't make sense to use this environment variable or nix path when I introduced buidling system by attrset in pr 320462, when I was inspired by nix-build arguments.

> Does `nixos-rebuild edit` not work for you?

yes, both `nixos-rebuild edit` and `nixos-rebuild edit --file /path/to/nixos/config/entry-point.nix` does not work.

rnhmjoj replied:
So, should I add a default /etc/nixos/configuration.nix? Or what?

amozeo replied:
if getting $NIXOS_CONFIG is empty and nix-config from nix-path fails, don't execute code where $NIXOS_CONFIG is used.

rnhmjoj replied:
Done.

amozeo replied:
not resolved, execution in line 225 can still fail
```suggestion
    NIXOS_CONFIG=${NIXOS_CONFIG:-$(runCmd nix-instantiate --find-file nixos-config 2> /dev/null || true)}
```

amozeo commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -218,6 +218,27 @@ runCmd() {
+    fi
+
+    # Try config.nixpkgs.source from NixOS config
+    if nixpkgsPath=$(runCmd nix-instantiate --eval --expr "((import <nixpkgs> {}).callPackage $NIXOS_CONFIG {}).nixpkgs.source" 2>/dev/null | jq -r); then return; fi
```
`pkgs.callPackage` of NixOS configuration module/file is not how you evaluate NixOS configuration. if someone has `nixpkgs.source` defined in different file, or under `config` attrname, this will fail.

```suggestion
    if nixpkgsPath=$(runCmd nix-instantiate --eval --expr "((import <nixpkgs/nixos/eval-config.nix> { modules = [ '$NIXOS_CONFIG' ]; }).config.nixpkgs.source" 2>/dev/null | jq -r); then return; fi
```
or
```suggestion
    if nixpkgsPath=$(runCmd nix-instantiate --eval --expr "((import <nixpkgs> { config = {}; overlays = []; }).nixos (import '$NIXOS_CONFIG')).config.nixpkgs.source" 2>/dev/null | jq -r); then return; fi
```

rnhmjoj replied:
Eh, you found the biggest hack of my proposal. Evaluating the whole NixOS configuration (with whatever `<nixpkgs>` is currently pointing to) is not generally possible: there may be errors due to newer options being used, or older being removed.

The only safe way is to use `pkgs.callPackage` to fill in whatever arguments the configuration.nix wants and evaluating *only* `nixpkgs.source` from the resuling attrset. This works as longs as users stick to the prescriptions I wrote in nixpkgs.source.description:

> In order to be evaluated before building the system, this option must be set in your configuration entry point (i.e. configuration.nix) and the value must not reference any of the NixOS module arguments (i.e. `config`, `pkgs`, etc.).

rnhmjoj replied:
I was wrong, I've found out about `_module.check = false` and for what I've tested so far it seems to be working. I can evaluate a configuration with errors and extract the `config.nixpkgs.source`.

rnhmjoj commented on nixos/modules/misc/nixpkgs-pin.nix:
```diff
@@ -6,10 +6,21 @@
 let
-  cfg = config.nixpkgs.flake;
+  cfg = config.nixpkgs;
+  usingFlakes = lib ? nixosSystem;
```
Probably yes. Currently the module could break if at some point `nixosSystem` were to be added to `lib` unconditionally, or renamed.
The option should be read-only and only set once in flake.nix.

rnhmjoj replied:
Done.

rnhmjoj committed:
nixos/top-level: add system.build.usingFlakes option

This option can be used by other NixOS modules to detect whether the
system is being built from a flake.

thiagokokada commented on pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/models.py:
```diff
@@ -57,7 +58,30 @@ def to_attr(self, *attrs: str) -> str:
+                    "nix-instantiate",
+                    "--eval",
+                    "--expr",
+                    f"""
+                      (import <nixpkgs/nixos> {{
+                        configuration = {{
+                          imports = [ {nixos_config} ];
+                          _module.check = false;
+                        }};
+                      }}).config.nixpkgs.source
+                    """,
```
Nit, but probably applying `textwrap.dedent` is a good idea here, otherwise it will unnecessary pass a very long string (and it will especially be bad in the logs if enabled).

rnhmjoj replied:
Sure, I had noticed the "\n                           \n                 " in the debug log. Done.

infinisil commented on pkgs/os-specific/linux/nixos-rebuild/nixos-rebuild.sh:
```diff
@@ -218,6 +218,34 @@ runCmd() {
+    fi
+
+    # Try config.nixpkgs.source from NixOS config
+    if nixpkgsPath=$(test -n "$NIXOS_CONFIG" && runCmd nix-instantiate --eval --expr "
+     (import <nixpkgs/nixos> {
+       configuration = {
+         imports = [ $NIXOS_CONFIG ];
+         _module.check = false;
+       };
+     }).config.nixpkgs.source
```
I'm really not a fan of this. It's fundamentally impossible to make `configuration.nix` specify Nixpkgs and it be used consistently, you always end up with the same problematic cases, such as:
- A separate Nixpkgs is needed to "bootstrap" the eval, like the `<nixpkgs>` here. And this is not reproducible in general.
- Upgrades across Nixpkgs versions are problematic, because some module might not evaluate with both the older and newer Nixpkgs (the `_module.check` can't ignore everything)

The real solution is to abandon the idea of "`configuration.nix` can reproduce the system" and switch to "`system.nix`/`flake.nix`/`default.nix` can reproduce the system", where that entry-point file depends on `configuration.nix`. That's something that Flakes got right, and it's also the direction I've indicated with https://github.com/NixOS/nixpkgs/pull/333788. This way there's none of the above problems.

RossComputerGuy replied:
Agreed, I kinda think of the non Flakes way as legacy and so large changes specific for it shouldn't be made. Flakes is the direction a lot of things are going and just manages it a lot cleaner.

infinisil replied:
I don't necessarily agree with that. I don't want people to have to buy into Flakes just to have an eval-reproducible system. It's very easy to make something similar work in traditional Nix, it's not a large change.

rnhmjoj replied:
> It's fundamentally impossible to make configuration.nix specify Nixpkgs and it be used consistently

My initial proposal was to use `callPackage` (or similar) to fill in the arguments of `configuration.nix` and extract `nixpkgs.source`, without using `evalModules`. This works reliably with the only limitation that `nixpkgs.source` must not depend on any of the module arguments and be set directly in `configuration.nix`.

> The real solution is to abandon the idea of "configuration.nix can reproduce the system"

Yes, but even then it's a letdown that you can't have a self-contained expression to describe a system reproducibly.
It means you'll never be able to just copy-paste, build and get the same system.

infinisil replied:
Nah it's totally possible. With https://github.com/NixOS/nixpkgs/pull/333788, you can write a `/etc/nixos/system.nix` as follows:

```nix
let
  nixpkgs = fetchTarball { ... };
in
import (nixpkgs + "/nixos") {
  configuration = {
    environment.systemPackages = [ ... ];
  };
}
```

And `nixos-rebuild switch` works.

nyabinary replied:
bump @rnhmjoj

rnhmjoj replied:
@nyabinary There's not much else to say as infinisil NAKed this idea.
From a UX point of view I think my solution is better, but the implementation is really hacky, so I'm ok with https://github.com/NixOS/nixpkgs/pull/333788 being chosen.
